tests/mir-opt/gvn.rs

```diff
@@ -816,6 +818,62 @@ fn casts_before_aggregate_raw_ptr(x: *const u32) -> *const [u8] {
     std::intrinsics::aggregate_raw_ptr(x, 4)
 }
 
+fn manual_slice_mut_len(x: &mut [i32]) -> usize {
+    // CHECK-LABEL: fn manual_slice_mut_len
+    // CHECK: _0 = PtrMetadata(_1);
+    let x: *mut [i32] = x;
+    let x: *const [i32] = x;
+    std::intrinsics::ptr_metadata(x)
+}
+
+// `.len()` on arrays ends up being something like this
+fn array_len(x: &mut [i32; 42]) -> usize {
+    // CHECK-LABEL: fn array_len
+    // CHECK: _0 = const 42_usize;
+    let x: &[i32] = x;
+    std::intrinsics::ptr_metadata(x)
+}
+
+#[custom_mir(dialect = "runtime")]
```

Comment:
This test is a bit tricky, could you add a small explanation why each case must work as shown?